File: src/iOS/FilterObjectsViewController.m
//
//  FilterObjectsViewController.m
//  Go Map!!
//
//  Created by Bryce Cogswell on 5/20/18.
//  Copyright © 2018 Bryce. All rights reserved.
//

#import "AppDelegate.h"
#import "EditorMapLayer.h"
#import "FilterObjectsViewController.h"
#import "MapView.h"

@interface FilterObjectsViewController ()
@end

@implementation FilterObjectsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	
	EditorMapLayer * editor = AppDelegate.shared.mapView.editorLayer;
	
	self.levelsText.text		= editor.showLevelRange;
	self.switchLevel.on			= editor.showLevel;
	self.switchPoints.on		= editor.showPoints;
	self.switchTrafficRoads.on	= editor.showTrafficRoads;
	self.switchServiceRoads.on	= editor.showServiceRoads;
	self.switchPaths.on			= editor.showPaths;
	self.switchBuildings.on		= editor.showBuildings;
	self.switchLanduse.on		= editor.showLanduse;
	self.switchBoundaries.on	= editor.showBoundaries;
	self.switchWater.on			= editor.showWater;
	self.switchRail.on			= editor.showRail;
	self.switchPower.on			= editor.showPower;
	self.switchPastFuture.on	= editor.showPastFuture;
	self.switchOthers.on		= editor.showOthers;
}

- (void)viewWillDisappear:(BOOL)animated
{
	[super viewWillDisappear:animated];

	EditorMapLayer * editor = AppDelegate.shared.mapView.editorLayer;
	
	editor.showLevelRange		= self.levelsText.text;
	editor.showLevel			= self.switchLevel.on;
	editor.showPoints			= self.switchPoints.on;
	editor.showTrafficRoads		= self.switchTrafficRoads.on;
	editor.showServiceRoads		= self.switchServiceRoads.on;
	editor.showPaths			= self.switchPaths.on;
	editor.showBuildings		= self.switchBuildings.on;
	editor.showLanduse			= self.switchLanduse.on;
	editor.showBoundaries		= self.switchBoundaries.on;
	editor.showWater			= self.switchWater.on;
	editor.showRail				= self.switchRail.on;
	editor.showPower			= self.switchPower.on;
	editor.showPastFuture		= self.switchPastFuture.on;
	editor.showOthers			= self.switchOthers.on;
}

// return a list of arrays, each array containing either a single integer or a first-last pair of integers
+(NSArray *)levelsForString:(NSString *)text
{
	NSMutableArray * list = [NSMutableArray new];
	NSScanner * scanner = [NSScanner scannerWithString:text];
	scanner.charactersToBeSkipped = [NSCharacterSet whitespaceAndNewlineCharacterSet];
	
	if ( [scanner isAtEnd] )
		return list;	// empty list
	for (;;) {
		NSInteger first, last;
		if ( ![scanner scanInteger:&first] )
			return nil;
		if ( [scanner scanString:@".." intoString:nil] ) {
			if ( ![scanner scanInteger:&last] )
				return nil;
			[list addObject:@[ @(first), @(last) ]];
		} else {
			[list addObject:@[ @(first) ]];
		}
		if ( [scanner isAtEnd] )
			return list;
		if ( ![scanner scanString:@"," intoString:nil] )
			return nil;
	}
}

- (void)setColorForText:(NSString *)text
{
	NSArray * a = [FilterObjectsViewController levelsForString:text];
	if ( a == nil ) {
		self.levelsText.textColor = UIColor.redColor;
	} else {
		self.levelsText.textColor = UIColor.blackColor;
	}
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
	NSString * newString = [textField.text stringByReplacingCharactersInRange:range withString:string];
	[self setColorForText:newString];
	return YES;
}

-(void)textFieldDidEndEditing:(UITextField *)textField
{
	[self setColorForText:textField.text];
}

@end

File: src/iOS/FilterObjectsViewController.swift
//
//  FilterObjectsViewController.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 5/20/18.
//  Copyright © 2018 Bryce. All rights reserved.
//

import UIKit

class FilterObjectsViewController: UITableViewController, UITextFieldDelegate {
    @IBOutlet weak var levelsText: UITextField!
    @IBOutlet weak var switchLevel: UISwitch!
    @IBOutlet weak var switchPoints: UISwitch!
    @IBOutlet weak var switchTrafficRoads: UISwitch!
    @IBOutlet weak var switchServiceRoads: UISwitch!
    @IBOutlet weak var switchPaths: UISwitch!
    @IBOutlet weak var switchBuildings: UISwitch!
    @IBOutlet weak var switchLanduse: UISwitch!
    @IBOutlet weak var switchBoundaries: UISwitch!
    @IBOutlet weak var switchWater: UISwitch!
    @IBOutlet weak var switchRail: UISwitch!
    @IBOutlet weak var switchPower: UISwitch!
    @IBOutlet weak var switchPastFuture: UISwitch!
    @IBOutlet weak var switchOthers: UISwitch!

    // return a list of arrays, each array containing either a single integer or a first-last pair of integers
	@objc
	class func levels(for text: String?) -> [[NSNumber]] {
		guard let text = text else { return [] }
		var list: [[NSNumber]] = []
		let scanner = Scanner(string: text)
        scanner.charactersToBeSkipped = CharacterSet.whitespacesAndNewlines

        if scanner.isAtEnd {
			return list // empty list
        }

        while true {
            var first = Int()
            var last = Int()
            if !scanner.scanInt(&first) {
                return []
			}
            if scanner.scanString("..", into: nil) {
                if !scanner.scanInt(&last) {
                    return []
                }
				list.append([NSNumber(value: first),NSNumber(value:last)])
			} else {
				list.append([NSNumber(value: first)])
			}
			if scanner.isAtEnd {
                return list
            }
            if !scanner.scanString(",", into: nil) {
				return []
            }
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let editor = AppDelegate.shared.mapView.editorLayer!

        levelsText.text = editor.showLevelRange
        switchLevel.isOn = editor.showLevel
		switchPoints.isOn = editor.showPoints
		switchTrafficRoads.isOn = editor.showTrafficRoads
		switchServiceRoads.isOn = editor.showServiceRoads
		switchPaths.isOn = editor.showPaths
		switchBuildings.isOn = editor.showBuildings
		switchLanduse.isOn = editor.showLanduse
		switchBoundaries.isOn = editor.showBoundaries
		switchWater.isOn = editor.showWater
		switchRail.isOn = editor.showRail
		switchPower.isOn = editor.showPower
		switchPastFuture.isOn = editor.showPastFuture
		switchOthers.isOn = editor.showOthers
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        let editor = AppDelegate.shared.mapView.editorLayer!

        editor.showLevelRange = levelsText.text!
        editor.showLevel = switchLevel.isOn
        editor.showPoints = switchPoints.isOn
        editor.showTrafficRoads = switchTrafficRoads.isOn
        editor.showServiceRoads = switchServiceRoads.isOn
        editor.showPaths = switchPaths.isOn
        editor.showBuildings = switchBuildings.isOn
        editor.showLanduse = switchLanduse.isOn
        editor.showBoundaries = switchBoundaries.isOn
        editor.showWater = switchWater.isOn
        editor.showRail = switchRail.isOn
        editor.showPower = switchPower.isOn
        editor.showPastFuture = switchPastFuture.isOn
        editor.showOthers = switchOthers.isOn
    }

	// show filter text in red if the level range is invalid
	func setColorForText(_ text: String?) {
        let a = FilterObjectsViewController.levels(for: text)
		if a.count == 0 {
			levelsText.textColor = UIColor.red
        } else {
            levelsText.textColor = UIColor.black
        }
    }

    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        let newString = (textField.text as NSString?)?.replacingCharacters(in: range, with: string)
        setColorForText(newString)
        return true
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        setColorForText(textField.text)
    }
}
